README.md
<img
  src="https://contrast.now.sh/fff/000?size=200&fontSize=2&baseline=2&fontWeight=900&radius=32&text=UIN"
  width="96"
  heigh="96"
/>

# Theme UI Native

Theme UI for React Native allows you to build consistent, theme-able React Native apps based on constraint-based design principles

[![GitHub][github-badge]][github]
[![Build Status][github-actions-badge]][github-actions]
[![Version][version]][npm]
![MIT License][license]
[![system-ui/theme][system-ui-badge]](https://system-ui.com/theme)
![][size]

[github]: https://github.com/samjbmason/theme-ui-native
[github-badge]: https://badgen.net/badge/-/github?icon=github&label
[github-actions]: https://github.com/samjbmason/theme-ui-native/actions
[github-actions-badge]: https://github.com/samjbmason/theme-ui-native/workflows/Run%20CI/badge.svg
[version]: https://badgen.net/npm/v/theme-ui-native
[npm]: https://npmjs.com/package/theme-ui-native
[license]: https://badgen.net/badge/license/MIT/blue
[system-ui-badge]: https://badgen.net/badge/system-ui/theme/black
[size]: https://badgen.net/bundlephobia/minzip/theme-ui-native

Built for React Native applications where customising colours, typography, and layout are treated as first-class citizens and based on a standard [Theme Specification](https://system-ui.com/theme), Theme UI For React Native is intended to work in a variety of applications and UI components. Colors, typography, and layout styles derived from customizable theme-based design scales help you build UI rooted in constraint-based design principles.

- [Getting started](#getting-started)
- [Usage](#usage)
- [Differences between Theme UI for Web & Theme UI for React Native](#differences-between-theme-ui-for-web-and-theme-ui-for-react-native)
- [API](#api)

## Getting started

```bash
npm install theme-ui-native
```

Any styles in your app can reference values from the global theme object. To provide the theme in context, wrap your application with the ThemeProvider component and pass in a custom theme object.

```jsx
// basic usage
import React from "react"
import { ThemeProvider } from "theme-ui-native"
import theme from "./theme"

export default props => (
  <ThemeProvider theme={theme}>{props.children}</ThemeProvider>
)
```

The theme object follows the System UI [Theme Specification](https://system-ui.com/theme), which lets you define custom color palettes, typographic scales, fonts, and more. Read more about [theming](https://theme-ui.com/theming/).

```jsx
// example theme.js
export default {
  fonts: {
    body: "Avenir Next",
    monospace: "Menlo, monospace"
  },
  colors: {
    text: "#000",
    background: "#fff",
    primary: "#33e"
  }
}
```

## Usage

### `sx` prop

The `sx` prop works similarly to Theme UI's `sx` prop, accepting style objects to add styles directly to an element in JSX, with theme-aware functionality. Using the `sx` prop for styles means that certain properties can reference values defined in your theme object. This is intended to make keeping styles consistent throughout your app the easy thing to do.

The `sx` prop only works in modules that have defined a custom pragma at the top of the file, which replaces the default `React.createElement` function. This means you can control which modules in your application opt into this feature without the need for a Babel plugin or additional configuration.

```jsx
/** @jsx jsx */
import { jsx } from "theme-ui-native"
import { Text } from "react-native"

export default props => (
  <Text
    sx={{
      fontWeight: "bold",
      fontSize: 4, // picks up value from `theme.fontSizes[4]`
      color: "primary" // picks up value from `theme.colors.primary`
    }}
  >
    Hello
  </Text>
)
```

### `styled` function

The `styled` function works similarly to Emotion's `styled` function, the first argument accepts a React component and the second argument accepts a style object that adds theme aware style properties to the `style` prop. The function returns a React component that can be used as normal within your application.

```jsx
import React from "react"
import { Text } from "react-native"
import { styled } from "theme-ui-native"

const Headline = styled(Text, {
  marginY: 2, // picks up value from `theme.space[2]`
  color: "primary" // picks up value from `theme.color.primary`
})

export default () => <Headline>Hello</Headline>
```

Components that are created with the styled function get the added ability of being able to use the `sx` prop. This allows you to set default styles and then override at each call site with theme aware values.

```jsx
import React from "react"
import { Text } from "react-native"
import { styled } from "theme-ui-native"

const Headline = styled(Text, {
  marginY: 2,
  color: "primary"
})

export default () => (
  <>
    <Headline>Hello</Headline> {/* output will use color `primary` */}
    <Headline
      sx={{
        color: "secondary" // output will use color `secondary`
      }}
    >
      Hello
    </Headline>
  </>
)
```

You can use a function as the second argument in `styled`, this function has access to all the props of the component and the theme object. This enables you to change values dynamically based on props.

```jsx
import React from "react"
import { Text } from "react-native"
import { styled } from "theme-ui-native"

const Headline = styled(Text, ({ isEnabled, theme }) => ({
  color: isEnabled ? "primary" : "secondary"
}))

export default () => <Headline isEnabled={true}>Hello</Headline>
```

### Raw values

If you would like to not use a theme value and instead use a literal value, you can pass the value as a string. The raw value is converted into a number so that it is compatible with React Natives style API.

```jsx
/** @jsx jsx */
import { jsx } from "theme-ui-native"
import { Text } from "react-native"

export default props => (
  <Text
    sx={{
      marginY: "2", // uses the raw value `2`
      marginX: 2 // picks up value from `theme.space[2]`
    }}
  >
    Hello
  </Text>
)
```

You can also use raw values by using the `style` prop as usual. These styles will take precedence over any styles created with the `sx` prop.

```jsx
/** @jsx jsx */
import { jsx } from "theme-ui-native"
import { Text } from "react-native"

export default props => (
  <Text
    sx={{
      color: 'primary'
      marginX: 2 // picks up value from `theme.space[2]`
    }}
    style={{color: '#000'}}
  >
    Hello
  </Text>
)
// Final output will be
// {color: '#000', marginX: 2}
```

### `sx` function

The `sx` function provides another option for adding theme aware style properties to your React components without the need for using the `styled` function or the `jsx` pragma.

```jsx

import { useTheme } from "theme-ui-native"
import { Text } from "react-native"

export default props => {
  const { sx } = useTheme()

  return (
    <Text
      style={sx({
        color: 'primary' // picks up value from `theme.colors.primary`
        marginX: 2 // picks up value from `theme.space[2]`
      })}
    >
      Hello
    </Text>
  )
}
// Final output will be
// {color: '#000', marginX: 2}
```

## Differences between Theme UI for Web and Theme UI for React Native

### Responsive styles

If you are coming from using Theme UI for web applications you are probably familiar with using arrays as values to change properties responsively based on breakpoints. This API isn't included Theme UI Native as there is currently no concept of responsive breakpoints within the React Native ecosystem.

### Body styles

There is no concept of the global styles within React Native and so this functionality is not available within Theme UI Native.

### MDX content

We haven't ported over MDX styling at this time as it seems unlikely to be used within the React Native context

### Color modes

We currently don't support color modes but are very open to the integrating them in future versions, feel free to raise a issue or PR if you have ideas on how to implement this.

## API

### `jsx`

The `jsx` export is a React create element function intended for use with a custom pragma comment. It adds support for the `sx` prop, which parses style objects with the Theme UI Native `css` utility.

```jsx
/** @jsx jsx */
import { jsx } from "theme-ui-native"
import { Text } from "react-native"

export default props => (
  <Text
    {...props}
    sx={{
      color: "primary"
    }}
  ></Text>
)
```

### `ThemeProvider`

The `ThemeProvider` provides context to components that use the `sx` prop.

| Prop       | Type   | Description            |
| ---------- | ------ | ---------------------- |
| `theme`    | Object | Theming context object |
| `children` | Node   | React children         |

### `styled`

The `styled` function allows you to create components that can be styled using theme aware styles. The newly created components also have access to the `sx` prop, which allows for per call site theme aware style overrides.

The first argument expects a react component, the second argument expects either a object containing styles or a function that returns a style object. If a function is used the props and theme are passed as the functions argument as an object.

```js
import { Text } from "react-native"

const Heading = styled(Text, { color: "primary" })

const Box = styled(View, ({ theme, ...props }) => ({
  color: theme.colors.primary
}))
```

### `useTheme`

The `useTheme` hook returns an object that contains full Theme UI Native context object, which includes the `theme` and the `sx` function.

```js
const { theme, sx } = useTheme()
```

### `sx`

The `sx` function is returned from the `useTheme` hook, you can use it to style your components with theme aware styles. You should use it within the `style` prop and it will return a style object with the theme values mapped to it. This works in a similar way to how the `sx` prop from the `jsx` pragma works under the hood.

```jsx
import React from "react"
import { View } from "react-native"
import { useTheme } from "theme-ui-native"

export default () => {
  const { sx } = useTheme()

  return <View style={sx({ mx: 2, color: "primary" })}>Hello world</View>
}
```

MIT License


CHANGELOG.md
# theme-ui-native

## 1.0.0

### Major Changes

- cbeaf46: `useTheme` hook now returns an object that contains the `theme` object and the `sx` function

  `sx` this function is returned from the `useTheme` hook and allows a user to add theme aware styles to a component without having to use `styled` or the `jsx` pragma. See README for more details.

  #### What

  The `useTheme` hook previously returned the theme object it now returns an object that contains `theme` prop and the `sx` prop

  #### Why

  We moved to an object as it allows for future features to be added into the `useTheme` hook without breaking changes

  #### How

  Previously you would access the theme object like so

  ```js
  const theme = useTheme()
  ```

  You will now need to access it like so

  ```js
  const { theme } = useTheme()
  ```


test/css.js
import { css, get } from "../src/css"

const theme = {
  colors: {
    primary: "tomato",
    secondary: "cyan",
    background: "white",
    text: "black"
  },
  fontSizes: [12, 14, 16, 24, 36],
  fonts: {
    monospace: "Menlo, monospace"
  },
  lineHeights: {
    body: 1.5
  },
  fontWeights: {
    bold: "600"
  },
  letterSpacings: ["-0.01em", "-0.02em", "0.01em"],
  space: [0, 4, 8, 16, 32, 64, 128, 256, "512"],
  sizes: {
    small: 4,
    medium: 8,
    large: 16,
    sidebar: 320
  },
  buttons: {
    primary: {
      p: 3,
      fontWeight: "bold",
      color: "white",
      bg: "primary",
      borderRadius: 2
    }
  },
  text: {
    caps: {
      fontSize: [1, 2],
      letterSpacing: "0.1em",
      textTransform: "uppercase"
    },
    title: {
      fontSize: [3, 4],
      letterSpacing: ["-0.01em", "-0.02em"]
    }
  },
  borderWidths: {
    thin: 1
  },
  radii: {
    small: 5
  }
}

describe("css", () => {
  test("returns an object", () => {
    const result = css()()
    expect(typeof result).toBe("object")
  })

  test("returns styles", () => {
    const result = css()({
      fontSize: 32,
      color: "blue",
      borderRadius: 4
    })
    expect(result).toEqual({
      fontSize: 32,
      color: "blue",
      borderRadius: 4
    })
  })

  test("handles all core styled system props", () => {
    const result = css({ theme })({
      m: 0,
      mb: 2,
      mx: "auto",
      p: 3,
      py: 4,
      fontSize: 3,
      fontWeight: "bold",
      color: "primary",
      bg: "secondary",
      fontFamily: "monospace",
      lineHeight: "body"
    })
    expect(result).toEqual({
      margin: 0,
      marginBottom: 8,
      marginHorizontal: "auto",
      padding: 16,
      paddingVertical: 32,
      color: "tomato",
      backgroundColor: "cyan",
      fontFamily: "Menlo, monospace",
      fontSize: 24,
      fontWeight: "600",
      lineHeight: 1.5
    })
  })

  test("works with functional arguments", () => {
    const result = css({ theme })(t => ({
      color: t.colors.primary
    }))
    expect(result).toEqual({
      color: "tomato"
    })
  })

  test("supports functional values", () => {
    const result = css({ theme })({
      color: t => t.colors.primary
    })
    expect(result).toEqual({
      color: "tomato"
    })
  })

  test("returns variants from theme", () => {
    const result = css(theme)({
      variant: "buttons.primary"
    })
    expect(result).toEqual({
      padding: 16,
      fontWeight: "600",
      color: "white",
      backgroundColor: "tomato",
      borderRadius: 2
    })
  })

  test("handles negative margins from scale", () => {
    const result = css(theme)({
      mt: -3,
      mx: -4
    })
    expect(result).toEqual({
      marginTop: -16,
      marginHorizontal: -32
    })
  })

  test("handles string negative values", () => {
    const result = css(theme)({
      letterSpacing: 1,
      mx: 4
    })
    expect(result).toEqual({
      letterSpacing: "-0.02em",
      marginHorizontal: 32
    })
  })

  test("handles negative string values from scale", () => {
    const result = css(theme)({
      my: -8,
      mx: 4
    })
    expect(result).toEqual({
      marginVertical: "-512",
      marginHorizontal: 32
    })
  })

  test("handles negative top, left, bottom, and right from scale", () => {
    const result = css(theme)({
      top: -1,
      right: -4,
      bottom: -3,
      left: -2
    })
    expect(result).toEqual({
      top: -4,
      right: -32,
      bottom: -16,
      left: -8
    })
  })

  test("multiples are transformed", () => {
    const style = css(theme)({
      marginX: 2,
      marginY: 2,
      paddingX: 2,
      paddingY: 2,
      size: "large"
    })
    expect(style).toEqual({
      marginHorizontal: 8,
      marginVertical: 8,
      paddingHorizontal: 8,
      paddingVertical: 8,
      width: 16,
      height: 16
    })
  })

  test("returns individual border styles", () => {
    const result = css(theme)({
      borderTopWidth: "thin",
      borderTopColor: "primary",
      borderTopLeftRadius: "small",
      borderTopRightRadius: "small",
      borderTopStartRadius: "small",
      borderTopEndRadius: "small",
      borderBottomWidth: "thin",
      borderBottomColor: "primary",
      borderBottomLeftRadius: "small",
      borderBottomRightRadius: "small",
      borderBottomStartRadius: "small",
      borderBottomEndRadius: "small",
      borderRightWidth: "thin",
      borderRightColor: "primary",
      borderLeftWidth: "thin",
      borderLeftColor: "primary"
    })
    expect(result).toEqual({
      borderTopColor: "tomato",
      borderTopWidth: 1,
      borderTopLeftRadius: 5,
      borderTopRightRadius: 5,
      borderTopStartRadius: 5,
      borderTopEndRadius: 5,
      borderBottomColor: "tomato",
      borderBottomWidth: 1,
      borderBottomLeftRadius: 5,
      borderBottomRightRadius: 5,
      borderBottomStartRadius: 5,
      borderBottomEndRadius: 5,
      borderRightColor: "tomato",
      borderRightWidth: 1,
      borderLeftColor: "tomato",
      borderLeftWidth: 1
    })
  })

  test("flexBasis uses theme.sizes", () => {
    const style = css(theme)({
      flexBasis: "sidebar"
    })
    expect(style).toEqual({
      flexBasis: 320
    })
  })

  test("string values are passed as raw value numbers", () => {
    const style = css(theme)({
      fontSize: "2",
      marginX: "auto",
      marginY: "4",
      margin: "4",
      paddingX: "2",
      paddingY: "2",
      padding: "2",
      borderTopWidth: "2",
      fontWeight: "600"
    })

    expect(style).toEqual({
      fontSize: 2,
      marginHorizontal: "auto",
      marginVertical: 4,
      margin: 4,
      paddingHorizontal: 2,
      paddingVertical: 2,
      padding: 2,
      borderTopWidth: 2,
      fontWeight: "600"
    })
  })

  test("handles style properties that are objects", () => {
    const style = css(theme)({
      textShadowOffset: { width: 1, height: 1 },
      color: "primary"
    })

    expect(style).toEqual({
      textShadowOffset: { width: 1, height: 1 },
      color: "tomato"
    })
  })

  test("handles non-scale values", () => {
    const result = css(theme)({
      textDecorationLine: "line-through",
      mx: 2
    })
    expect(result).toEqual({
      textDecorationLine: "line-through",
      marginHorizontal: 8
    })
  })

  test("uses raw value if value outside of scale is used", () => {
    const result = css(theme)({
      fontSize: 10,
      mx: 10
    })
    expect(result).toEqual({
      fontSize: 10,
      marginHorizontal: 10
    })
  })
})

describe("get", () => {
  test("fallbacks to default value if object isnt present", () => {
    const result = get(null, 1, {})
    expect(result).toEqual({})
  })
})


test/index.js
import React from "react"
import { Text } from "react-native"
import { render } from "@testing-library/react-native"
import { toHaveStyle, toHaveProp } from "@testing-library/jest-native"
import { ThemeProvider, useTheme, jsx, styled, sx } from "../src"

expect.extend({ toHaveStyle, toHaveProp })

describe("ThemeProvider", () => {
  test("renders", () => {
    const { container } = render(
      <ThemeProvider>
        <Text>Hello</Text>
      </ThemeProvider>
    )
    expect(container).toMatchSnapshot()
  })
})

describe("useTheme", () => {
  test("returns theme context", () => {
    let context
    const GetContext = () => {
      context = useTheme()
      return false
    }
    render(
      <ThemeProvider
        theme={{
          colors: {
            text: "tomato"
          }
        }}
      >
        <GetContext />
      </ThemeProvider>
    )

    expect(context).toBeTruthy()
    expect(context.theme.colors.text).toBe("tomato")
  })

  test("returns sx function", () => {
    let context
    const GetContext = () => {
      context = useTheme()
      return false
    }
    render(
      <ThemeProvider
        theme={{
          colors: {
            text: "tomato"
          }
        }}
      >
        <GetContext />
      </ThemeProvider>
    )

    expect(context).toBeTruthy()
    expect(context.sx).toBeInstanceOf(Function)
  })
})

describe("jsx", () => {
  test("custom pragma adds styles", () => {
    const { getByTestId } = render(
      jsx(Text, {
        sx: { marginHorizontal: 2, fontSize: 2, color: "red" },
        testID: "test"
      })
    )

    expect(getByTestId("test")).toHaveStyle({
      marginHorizontal: 8
    })
    expect(getByTestId("test")).toHaveStyle({
      fontSize: 16
    })
    expect(getByTestId("test")).toHaveStyle({
      color: "red"
    })
  })

  test("custom pragma handles null props", () => {
    const json = render(jsx(Text, null, "hello"))
    expect(json).toMatchSnapshot()
  })

  test("removes sx prop", () => {
    const { getByTestId } = render(
      jsx(Text, {
        sx: { marginHorizontal: 2, fontSize: 2, color: "red" },
        testID: "test"
      })
    )
    expect(getByTestId("test")).not.toHaveProp("sx")
  })

  test("adds raw values from style prop", () => {
    const { getByTestId } = render(
      jsx(Text, {
        style: { marginHorizontal: 2 },
        testID: "test"
      })
    )

    expect(getByTestId("test")).toHaveStyle({ marginHorizontal: 2 })
  })

  test("sx prop can accept a function", () => {
    const { getByTestId } = render(
      jsx(
        ThemeProvider,
        {
          theme: {
            colors: {
              primary: "tomato"
            }
          }
        },
        jsx(Text, {
          sx: t => ({ color: t.colors.primary }),
          testID: "test"
        })
      )
    )
    expect(getByTestId("test")).toHaveStyle({ color: "tomato" })
  })

  test("sx and style prop can be used together", () => {
    const { getByTestId } = render(
      jsx(Text, {
        style: {
          margin: 0
        },
        sx: {
          backgroundColor: "tomato"
        },
        testID: "test"
      })
    )
    expect(getByTestId("test")).toHaveStyle({ backgroundColor: "tomato" })
    expect(getByTestId("test")).toHaveStyle({ margin: 0 })
  })

  test("sx prop supports dot notation", () => {
    const { getByTestId } = render(
      jsx(
        ThemeProvider,
        {
          theme: {
            colors: {
              text: "black",
              base: {
                blue: ["#07c"],
                primary: "cyan"
              }
            }
          }
        },
        jsx(Text, {
          sx: {
            color: "base.blue.0",
            backgroundColor: "base.primary"
          },
          testID: "test"
        })
      )
    )
    expect(getByTestId("test")).toHaveStyle({ backgroundColor: "cyan" })
    expect(getByTestId("test")).toHaveStyle({ color: "#07c" })
  })

  test("does not add style prop when not provided", () => {
    const { getByTestId } = render(jsx(Text, { testID: "test" }, "hi"))
    expect(getByTestId("test")).not.toHaveProp("style")
  })
})

describe("styled", () => {
  test("does not add style prop when not provided", () => {
    const Test = styled(Text)
    const { getByTestId } = render(<Test testID="test" />)

    expect(getByTestId("test")).not.toHaveProp("style")
  })

  test("works when no styles are added", () => {
    const Test = styled(Text, {})
    const { container } = render(<Test />)
    expect(container).toMatchSnapshot()
  })

  test("adds styles when using function call argument", () => {
    const theme = { colors: { primary: "#07c" }, space: [0, 8, 16] }
    const Test = styled(Text, {
      marginHorizontal: 2,
      color: "tomato",
      backgroundColor: "primary"
    })

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Test testID="test" />
      </ThemeProvider>
    )

    expect(getByTestId("test")).toHaveStyle({ backgroundColor: "#07c" })
    expect(getByTestId("test")).toHaveStyle({ color: "tomato" })
    expect(getByTestId("test")).toHaveStyle({ marginHorizontal: 16 })
  })

  test("adds styles when passing function to argument", () => {
    const theme = { colors: { primary: "#07c" }, space: [0, 8, 16] }
    const Test = styled(Text, ({ marginH }) => ({
      marginHorizontal: marginH,
      color: "tomato",
      backgroundColor: "primary"
    }))

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Test marginH={2} testID="test" />
      </ThemeProvider>
    )

    expect(getByTestId("test")).toHaveStyle({ backgroundColor: "#07c" })
    expect(getByTestId("test")).toHaveStyle({ color: "tomato" })
    expect(getByTestId("test")).toHaveStyle({ marginHorizontal: 16 })
  })

  test("adds styles when using sx prop", () => {
    const theme = { colors: { primary: "#07c" }, space: [0, 8, 16] }
    const Test = styled(Text)

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Test
          testID="test"
          sx={{
            marginHorizontal: 2,
            color: "tomato",
            backgroundColor: "primary"
          }}
        />
      </ThemeProvider>
    )

    expect(getByTestId("test")).toHaveStyle({ backgroundColor: "#07c" })
    expect(getByTestId("test")).toHaveStyle({ color: "tomato" })
    expect(getByTestId("test")).toHaveStyle({ marginHorizontal: 16 })
  })

  test("styles passed in through sx prop override argument styles", () => {
    const Test = styled(Text, { marginHorizontal: 16, color: "red" })
    const { getByTestId } = render(
      <Test sx={{ marginHorizontal: 20 }} testID="test" />
    )
    expect(getByTestId("test")).toHaveStyle({ marginHorizontal: 20 })
    expect(getByTestId("test")).toHaveStyle({ color: "red" })
  })

  test("works with style prop, sx prop and argument styles", () => {
    const theme = { colors: { primary: "#07c" }, space: [0, 8, 16] }
    const Test = styled(Text, { marginHorizontal: 0, color: "red" })
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Test
          style={{ marginHorizontal: 30 }}
          sx={{ marginHorizontal: 2, color: "primary" }}
          testID="test"
        />
      </ThemeProvider>
    )
    expect(getByTestId("test")).toHaveStyle({ marginHorizontal: 30 })
    expect(getByTestId("test")).toHaveStyle({ color: "#07c" })
  })
})

describe("sx", () => {
  test("adds style object to component", () => {
    const theme = { colors: { primary: "#07c" }, space: [0, 8, 16] }

    const Test = () => {
      const { sx } = useTheme()
      return (
        <Text
          testID="test"
          style={sx({
            marginHorizontal: 2,
            color: "tomato",
            backgroundColor: "primary"
          })}
        />
      )
    }

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Test />
      </ThemeProvider>
    )

    expect(getByTestId("test")).toHaveStyle({ backgroundColor: "#07c" })
    expect(getByTestId("test")).toHaveStyle({ color: "tomato" })
    expect(getByTestId("test")).toHaveStyle({ marginHorizontal: 16 })
  })
})


src/index.js
import { createContext, createElement, useContext } from "react"
import { css } from "./css"

const ThemeContext = createContext({})

export const useTheme = () => {
  const ctx = useContext(ThemeContext)
  return { ...ctx, sx: css(ctx.theme) }
}

const getStyles = (props, theme) => {
  if (!props.sx && !props.style) return undefined
  const styles = typeof props.sx === "function" ? props.sx(theme) : props.sx
  const parsedStyles = css(theme)(styles)
  const raw = props.style
  return { ...parsedStyles, ...raw }
}

const parseProps = (props, theme) => {
  if (!props) return null
  const next = {}
  for (let key in props) {
    if (key === "sx") continue
    next[key] = props[key]
  }

  const styles = getStyles(props, theme)
  if (styles) next.style = styles
  return next
}

const hasStyles = obj => Object.getOwnPropertyNames(obj).length > 0

export const ThemeProvider = ({ theme, children }) =>
  createElement(ThemeContext.Provider, { value: { theme } }, children)

export const jsx = (type, props, ...children) => {
  return createElement(ThemeContext.Consumer, {}, ({ theme }) => {
    return createElement(type, { ...parseProps(props, theme) }, ...children)
  })
}

export const styled = (type, style) => props => {
  const { theme } = useTheme()
  const styles =
    typeof style === "function" ? style({ ...props, theme }) : style
  const sx = { ...styles, ...props.sx }
  const mergedProps = { ...props, ...(hasStyles(sx) && { sx }) }

  const parsedProps = parseProps(mergedProps, theme)

  return createElement(type, parsedProps)
}


src/css.js
const defaultTheme = {
  space: [0, 4, 8, 16, 32, 64, 128, 256, 512],
  fontSizes: [12, 14, 16, 20, 24, 32, 48, 64, 72]
}

export const get = (obj, key, def, p, undef) => {
  const path = key && typeof key === "string" ? key.split(".") : [key]
  for (p = 0; p < path.length; p++) {
    obj = obj ? obj[path[p]] : undef
  }
  return obj === undef ? def : obj
}

const aliases = {
  bg: "backgroundColor",
  m: "margin",
  mt: "marginTop",
  mr: "marginRight",
  mb: "marginBottom",
  ml: "marginLeft",
  mx: "marginHorizontal",
  my: "marginVertical",
  marginX: "marginHorizontal",
  marginY: "marginVertical",
  p: "padding",
  pt: "paddingTop",
  pr: "paddingRight",
  pb: "paddingBottom",
  pl: "paddingLeft",
  px: "paddingHorizontal",
  py: "paddingVertical",
  paddingX: "paddingHorizontal",
  paddingY: "paddingVertical"
}

const multiples = {
  size: ["width", "height"]
}

export const scales = {
  width: "sizes",
  height: "sizes",
  bottom: "space",
  end: "space",
  left: "space",
  right: "space",
  start: "space",
  top: "space",
  minWidth: "sizes",
  maxWidth: "sizes",
  minHeight: "sizes",
  maxHeight: "sizes",
  margin: "space",
  marginTop: "space",
  marginRight: "space",
  marginBottom: "space",
  marginLeft: "space",
  marginStart: "space",
  marginEnd: "space",
  marginHorizontal: "space",
  marginVertical: "space",
  padding: "space",
  paddingTop: "space",
  paddingRight: "space",
  paddingBottom: "space",
  paddingLeft: "space",
  paddingStart: "space",
  paddingEnd: "space",
  paddingHorizontal: "space",
  paddingVertical: "space",
  paddingX: "space",
  paddingY: "space",
  borderWidth: "borderWidths",
  borderTopWidth: "borderWidths",
  borderRightWidth: "borderWidths",
  borderBottomWidth: "borderWidths",
  borderLeftWidth: "borderWidths",
  borderEndWidth: "borderWidths",
  borderStartWidth: "borderWidths",
  zIndex: "zIndices",
  shadowColor: "colors",
  backgroundColor: "colors",
  borderColor: "colors",
  borderBottomColor: "colors",
  borderEndColor: "colors",
  borderLeftColor: "colors",
  borderRightColor: "colors",
  borderStartColor: "colors",
  borderTopColor: "colors",
  borderRadius: "radii",
  borderBottomEndRadius: "radii",
  borderBottomLeftRadius: "radii",
  borderBottomRightRadius: "radii",
  borderBottomStartRadius: "radii",
  borderTopEndRadius: "radii",
  borderTopLeftRadius: "radii",
  borderTopRightRadius: "radii",
  borderTopStartRadius: "radii",
  color: "colors",
  fontFamily: "fonts",
  fontSize: "fontSizes",
  fontWeight: "fontWeights",
  textShadowColor: "colors",
  letterSpacing: "letterSpacings",
  lineHeight: "lineHeights",
  textDecorationColor: "colors",
  tintColor: "colors",
  overlayColor: "colors",
  flexBasis: "sizes",
  size: "sizes"
}

const positiveOrNegative = (scale, value) => {
  if (typeof value !== "number" || value >= 0) {
    return get(scale, value, value)
  }
  const absolute = Math.abs(value)
  const n = get(scale, absolute, absolute)
  if (typeof n === "string") return `-${n}`
  return Number(n) * -1
}

const transforms = [
  "marginTop",
  "marginRight",
  "marginBottom",
  "marginLeft",
  "marginStart",
  "marginEnd",
  "marginHorizontal",
  "marginVertical",
  "top",
  "bottom",
  "left",
  "right",
  "start",
  "end"
].reduce(
  (acc, curr) => ({
    ...acc,
    [curr]: positiveOrNegative
  }),
  {}
)

const maintainStringType = ["fontWeight"]

export const css = (props = {}) => (args = {}) => {
  const theme = {
    ...defaultTheme,
    ...("theme" in props ? props.theme : props)
  }
  let result = {}
  const styles = typeof args === "function" ? args(theme) : args

  for (const key in styles) {
    const x = styles[key]
    const val = typeof x === "function" ? x(theme) : x

    if (key === "variant") {
      const variant = css(theme)(get(theme, val))
      result = { ...result, ...variant }
      continue
    }

    if (val && typeof val === "object") {
      result[key] = css(theme)(val)
      continue
    }

    const prop = key in aliases ? aliases[key] : key

    if (
      !maintainStringType.includes(prop) &&
      typeof val === "string" &&
      !isNaN(val)
    ) {
      result[prop] = Number(val)
      continue
    }

    const scaleName = prop in scales ? scales[prop] : undefined
    const scale = get(theme, scaleName, get(theme, prop, {}))
    const transform = get(transforms, prop, get)
    const value = transform(scale, val, val)

    if (multiples[prop]) {
      const dirs = multiples[prop]
      for (let i = 0; i < dirs.length; i++) {
        result[dirs[i]] = value
      }
    } else {
      result[prop] = value
    }
  }

  return result
}
